### PR TITLE
Concordia'/issuer frontends

### LINKED ISSUE #63: [Concordia] Create independent frontends for each issuer
For convenience, issuing in concordia is currently done through the verifier UI. To provide an example of how to create an issuer in general, it would be helpful to have frontends for each issuer included in the issuer project.

**Sub tasks**
- [ ] Setup telegram bot for telegram login
- [x] Implement simple issuer frontends
- [x] Configure issuer servers to serve corresponding frontend

### PR DESCRIPTION
**Pull request by soerenbf**
## Purpose

Provide separate frontends for SoMe issuer services

Closes #63 

## Changes

- Added frontend project for some-issuer example
- Configured services to serve frontends

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: examples/some-issuer/frontend/src/scss/_bootstrap-config.scss [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+// @import 'bootstrap/scss/list-group';
+// @import 'bootstrap/scss/close';
+// @import "bootstrap/scss/toasts";
+// @import 'bootstrap/scss/modal';
+// @import "bootstrap/scss/tooltip";
+// @import "bootstrap/scss/popover";
+// @import "bootstrap/scss/carousel";
+// @import "bootstrap/scss/spinners";
+// @import "bootstrap/scss/offcanvas";
+// @import "bootstrap/scss/placeholders";
```
Comment (abizjak) [inline examples/some-issuer/frontend/src/scss/_bootstrap-config.scss:39]: Do we want to keep this?

File: examples/some-issuer/frontend/src/shared/Layout/Layout.tsx [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        <main className="layout">
+            <div className="d-inline-flex align-items-center">
+                <div className="layout__logo">
+                    <img src={ccdLogo} alt="Concordium logo" />
+                </div>
+                <div className="layout__plus">+</div>
+                <div className="layout__logo">{logo}</div>
+            </div>
+            <h1>{platform} web3 ID issuer</h1>
+            <h4 className="mb-4">Create your web3 ID credential for {platform} by logging in</h4>
```
Comment (abizjak) [inline examples/some-issuer/frontend/src/shared/Layout/Layout.tsx:20]:
```suggestion
            <h4 className="mb-4">Create your Web3 ID credential for {platform} by logging in using</h4>
```

Comment (abizjak) [reply examples/some-issuer/frontend/src/shared/Layout/Layout.tsx:20]: Could we elaborate here a bit, telling the user what will happen more explicitly.

File: examples/some-issuer/frontend/src/shared/Layout/Layout.tsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    return (
+        <main className="layout">
+            <div className="d-inline-flex align-items-center">
+                <div className="layout__logo">
+                    <img src={ccdLogo} alt="Concordium logo" />
+                </div>
+                <div className="layout__plus">+</div>
+                <div className="layout__logo">{logo}</div>
+            </div>
+            <h1>{platform} web3 ID issuer</h1>
```
Comment (abizjak) [inline examples/some-issuer/frontend/src/shared/Layout/Layout.tsx:19]:
```suggestion
            <h1>{platform} Web3 ID issuer</h1>
```

File: examples/some-issuer/frontend/src/shared/util.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+        if (!response.ok) throw new Error('Error getting credential: ' + (await response.text()));
+
+        const { txHash: hash, credential } = (await response.json()) as IssuerResponse;
+        txHash = hash;
+        const { proof, randomness } = credential;
+        return { proof, randomness };
+    });
+
+    console.log('Transaction submitted, hash:', txHash);
```
Comment (abizjak) [inline examples/some-issuer/frontend/src/shared/util.ts:104]: Do we want to keep this?

File: examples/some-issuer/frontend/src/shared/util.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        if (req.platform === Platform.Telegram) body.telegramUser = req.user;
+
+        const endpoint = ISSUER_URL + 'credential';
+        const response = await fetch(endpoint, {
+            method: 'POST',
+            headers: {
+                'Content-Type': 'application/json',
+            },
+            // If we're sending Discord requests, we need to include proof of our user id
+            credentials: 'include',
```
Comment (abizjak) [inline examples/some-issuer/frontend/src/shared/util.ts:92]: Should this then always be enabled, even for telegram?

File: examples/some-issuer/frontend/src/shared/util.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    while (true) {
+        try {
+            const itemSummary = await api.getGrpcClient().waitForTransactionFinalization(txHash!);
+            console.log('Transaction completed.', itemSummary);
+            break;
+        } catch (error) {
+            // NOT_FOUND errors just mean that the transaction hasn't been propagated yet
+            if ((error as RpcError).code !== 'NOT_FOUND') throw error;
+            // Sleep for half a second and try again
+            console.log('Transaction not found. Retrying in 200ms...');
```
Comment (abizjak) [inline examples/some-issuer/frontend/src/shared/util.ts:117]: These comments and the code don't match.

File: examples/some-issuer/frontend/src/shared/util.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    });
+
+    console.log('Transaction submitted, hash:', txHash);
+
+    // Loop until transaction has been finalized
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+        try {
+            const itemSummary = await api.getGrpcClient().waitForTransactionFinalization(txHash!);
+            console.log('Transaction completed.', itemSummary);
```
Comment (abizjak) [inline examples/some-issuer/frontend/src/shared/util.ts:111]: Could we not console.log in production?

File: examples/some-issuer/src/bin/discord.rs [hunk 8]
```diff
@@ -23,13 +23,14 @@ use http::{HeaderValue, StatusCode};
...
+use serde_json::json;
 use some_issuer::{set_shutdown, IssueResponse, IssuerState};
-use std::{net::SocketAddr, path::PathBuf, sync::Arc, time::Duration};
+use std::{fs, net::SocketAddr, path::PathBuf, sync::Arc, time::Duration};
 use tonic::transport::ClientTlsConfig;
 use tower_http::{cors::CorsLayer, services::ServeDir};
 
 const DISCORD_API_ENDPOINT: &str = "https://discord.com/api/v10";
-
+const HTML_TITLE: &str = "Discord web3 ID issuer";
```
Comment (abizjak) [inline examples/some-issuer/src/bin/discord.rs:33]:
```suggestion
const HTML_TITLE: &str = "Discord Web3 ID issuer";
```

File: examples/some-issuer/src/bin/telegram.rs [hunk 9]
```diff
@@ -8,16 +13,20 @@ use concordium_rust_sdk::{
...
+use serde::{Deserialize, Serialize};
+use serde_json::json;
 use sha2::{Digest, Sha256};
 use some_issuer::{set_shutdown, IssueResponse, IssuerState};
-use std::{fmt::Write, net::SocketAddr, path::PathBuf, sync::Arc};
+use std::{fmt::Write, fs, net::SocketAddr, path::PathBuf, sync::Arc};
 use tonic::transport::ClientTlsConfig;
 use tower_http::{cors::CorsLayer, services::ServeDir};
 
+const HTML_TITLE: &str = "Telegram web3 ID issuer";
```
Comment (abizjak) [inline examples/some-issuer/src/bin/telegram.rs:28]:
```suggestion
const HTML_TITLE: &str = "Telegram Web3 ID issuer";
```

File: examples/some-issuer/templates/discord-oauth.hbs [hunk 10]
```diff
@@ -10,7 +10,8 @@
         username: '{{username}}',
         state,
       };
-      window.opener.postMessage(message, '{{dapp_domain}}/');
+      window.opener.postMessage(message, '{{dapp_domain}}');
+      window.opener.postMessage(message, '{{verifier_dapp_domain}}');
```
Comment (abizjak) [inline examples/some-issuer/templates/discord-oauth.hbs:14]: Is it correct that only one of these will actually send the message, depending on whether the popup originated from the issuer or verifier frontend?

File: examples/some-issuer/frontend/README.md [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+
+In the project directory, you can run:
+
+### `yarn build-[telegram|discord]`
+
+Builds the app for production to the `dist/[telegram|discord]` folder.
+
+### `yarn dev-[telegram|discord]`
+
+Runs the frontend for the corresponding issuer in the development mode.
```
Comment (abizjak) [inline examples/some-issuer/frontend/README.md:13]: We should explain here that you need to configure some things for this to actually work.

It'd be good to list the environment variables that enable this.

File: examples/some-issuer/frontend/src/scss/_bootstrap-config.scss (hunk 1 above)
Comment (soerenbf) [reply examples/some-issuer/frontend/src/scss/_bootstrap-config.scss:39]: I'll remove them with a comment that they can be added if the corresponding components are added

File: examples/some-issuer/frontend/src/scss/_layout.scss [hunk 12]
```diff
@@ -0,0 +1,5 @@
+// Project layout
+@import 'layout/root';
+// @import 'layout/btn';
+// @import 'layout/modal';
+// @import 'layout/accordion';
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply examples/some-issuer/frontend/src/scss/_layout.scss:5]: same here

File: examples/some-issuer/frontend/src/shared/util.ts (hunk 4 above)
Comment (soerenbf) [reply examples/some-issuer/frontend/src/shared/util.ts:104]: Well.. it's good for debugging, and it doesn't hurt much does it?

Comment (abizjak) [reply examples/some-issuer/frontend/src/shared/util.ts:104]: I think it does. Any time I open a console on the page and see all kinds of debug logs I think people don't care about quality.

File: examples/some-issuer/templates/discord-oauth.hbs (hunk 10 above)
Comment (soerenbf) [reply examples/some-issuer/templates/discord-oauth.hbs:14]: Both are tried every time. It results in an error in the console, as one of them don't match. Can see if I can find another way around, but it works fine it its current state.

File: examples/some-issuer/frontend/README.md (hunk 11 above)
Comment (soerenbf) [reply examples/some-issuer/frontend/README.md:13]: Yes, good point. Will add this

File: examples/some-issuer/frontend/src/shared/util.ts (hunk 5 above)
Comment (soerenbf) [reply examples/some-issuer/frontend/src/shared/util.ts:92]: I don't know, this is how the request is done in the issuer as well, and I just copied the implementation in the interest of time.

File: examples/some-issuer/frontend/src/shared/util.ts (hunk 4 above)
Comment (abizjak) [reply examples/some-issuer/frontend/src/shared/util.ts:104]: Could at least do `console.debug` and then you only see it if you care to enable `verbose` mode.

Comment (soerenbf) [reply examples/some-issuer/frontend/src/shared/util.ts]: I added something that provides better user feedback instead.

**Pull request merged**